File: data-structures/stacks/PHP/Stack.php
<?php

include '../../../helpers/PHP/helpers.php';

class StackNode
{
    public $data;
    public $next;

    public function __construct($data, $next)
    {
        $this->data = $data;
        $this->next = $next;
    }
}

class Stack
{
    public $top;

    public function push($data)
    {
        $node = new StackNode($data, $this->top);
        $this->top = $node;
    }

    public function pop()
    {
        if ($this->top === null) {
            return null;
        }

        $top = $this->top->data;
        $this->top = $this->top->next;
        return $top;
    }

    public function top()
    {
        if ($this->top === null) {
            return null;
        }
        
        return $this->top->data;
    }

    public function isEmpty()
    {
        return $this->top === null;
    }

    public function display()
    {
        $node = $this->top;
        $count = 0;

        if ($node !== null) {
            while ($node !== null) {
                println('- Node data: ' . $node->data);
                $node = $node->next;
                $count++;
            }
        }
        println('-- Node count: ' . $count);
        println('');
    }
}

// Create stack
$stack = new Stack();
$stack->display();
stackAssertEqual('', $stack);

// Push
println("---------- Push ----------");
$stack->push(42);
$stack->display();
stackAssertEqual('42', $stack);

$stack->push(51);
$stack->display();
stackAssertEqual('51,42', $stack);

$stack->push(11);
$stack->push(455);
$stack->display();
stackAssertEqual('455,11,51,42', $stack);

// Pop
println("---------- Pop ----------");
$stack = new Stack();
$stack->push(42);
$stack->push(51);
$stack->push(11);
$stack->push(455);

$pop = $stack->pop();
stackAssertEqual('11,51,42', $stack);
assertEqual(455, $pop);

$pop = $stack->pop();
stackAssertEqual('51,42', $stack);
assertEqual(11, $pop);

$pop = $stack->pop();
assertEqual(51, $pop);

$pop = $stack->pop();
assertEqual(42, $pop);

$pop = $stack->pop();
assertEqual(null, $pop);

// Top
println("---------- Top ----------");
$stack = new Stack();
$stack->push(42);
$stack->push(51);

$top = $stack->top();
stackAssertEqual('51,42', $stack);
assertEqual(51, $top);

// Is empty
println("---------- Is empty ----------");
$stack = new Stack();
assertEqual(true, $stack->isEmpty());

$stack->push(42);
assertEqual(false, $stack->isEmpty());

File: data-structures/stacks/Java/Stack.java
public class Stack {
    private class StackNode {
        private int data;
        private StackNode next;

        private StackNode(int data, StackNode next) {
            this.data = data;
            this.next = next;
        }
    }

    private StackNode top;

    public void push(int data) {
        StackNode node = new StackNode(data, top);
        top = node;
    }

    public int pop() {
        if (top == null) {
            return -1;
        }

        int data = top.data;
        top = top.next;
        return data;
    }

    public int top() {
        if (top == null) {
            return -1;
        }

        return top.data;
    }

    public boolean isEmpty() {
        return top == null;
    }

    public void display() {
        if (top == null) {
            System.out.println("[]");
            return;
        }

        StackNode current = top;
        System.out.println("[");
        while (current != null) {
            System.out.println("\t" + current.data);
            current = current.next;
        }
        System.out.println("]");
    } 

    public static void main(String[] args) {
        // Create stack
        Stack stack = new Stack();
        stack.display();

        // Push
        System.out.println("---------- Push ----------");
        stack.push(42);
        stack.display();

        stack.push(51);
        stack.display();

        stack.push(11);
        stack.push(455);
        stack.display();

        // Pop
        System.out.println("---------- Pop ----------");
        stack = new Stack();
        int pop = stack.pop();
        System.out.println("pop: " + pop);

        stack.push(42);
        stack.push(51);
        stack.push(11);
        stack.push(455);

        pop = stack.pop();
        System.out.println("pop: " + pop);

        pop = stack.pop();
        System.out.println("pop: " + pop);

        pop = stack.pop();
        System.out.println("pop: " + pop);

        pop = stack.pop();
        System.out.println("pop: " + pop);

        pop = stack.pop();
        System.out.println("pop: " + pop);

        // Top
        System.out.println("---------- Top ----------");
        stack = new Stack();
        int top = stack.top();
        System.out.println("top: " + top);

        stack.push(42);
        stack.push(51);

        top = stack.top();
        System.out.println("top: " + top);

        // Is empty
        System.out.println("---------- Is empty ----------");
        stack = new Stack();
        System.out.println(stack.isEmpty());

        stack.push(42);
        System.out.println(stack.isEmpty());
    }
}